File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/ShaderProgram.java
package com.darkraha.opengldemoj.gl;


import com.darkraha.opengldemoj.gl.shader.ShaderProgramBuilder;
import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.lwjgl.system.MemoryUtil;

import java.nio.FloatBuffer;

import static com.darkraha.opengldemoj.gl.shader.ShaderProgramBuilder.*;
import static org.lwjgl.opengl.GL20.glGetUniformLocation;
import static org.lwjgl.opengl.GL33.*;

public class ShaderProgram {
    public static final FloatBuffer MATRIX_BUFFER = MemoryUtil.memAllocFloat(16);

    private int[] ids;
    private final int[] matricesLocations = new int[U_MATRIX_NAMES.length];
    private final int[] samplersLocations = new int[8];
    private int[] light = new int[3];
    private int normalLocation;

    public int solidColorLocation;
    public int idProgram;


    public ShaderProgram() {
        ShaderProgramBuilder builder = new ShaderProgramBuilder();
        builder.colors(false,true).matrix();
        init(builder.buildVertexShader(), builder.buildFragmentShader());
    }

    public ShaderProgram(String vertexShaderCode, String fragmentShaderCode) {
        init(vertexShaderCode, fragmentShaderCode);
    }

    protected void init(String vertexShaderCode, String fragmentShaderCode) {
        ids = createProgram(vertexShaderCode, fragmentShaderCode);
        idProgram = ids[0];
        setupLocations();
    }

    public void setupLocations() {
        glUseProgram(idProgram);

        for (int i = 0; i < matricesLocations.length; ++i) {
            matricesLocations[i] = glGetUniformLocation(idProgram, U_MATRIX_NAMES[i]);
        }

        samplersLocations[0] = glGetUniformLocation(idProgram, U_SAMPLER_NAME);

        for (int i = 1; i < samplersLocations.length; ++i) {
            samplersLocations[i] = glGetUniformLocation(idProgram, U_SAMPLER_NAME + i);
        }

        solidColorLocation = glGetUniformLocation(idProgram, U_SOLID_COLOR_NAME);
        light[0] = glGetUniformLocation(idProgram, U_LIGHT_NAMES[0]);
        light[1] = glGetUniformLocation(idProgram, U_LIGHT_NAMES[1]);
        light[2] = glGetUniformLocation(idProgram, U_LIGHT_NAMES[2]);
    }


    public void use() {
        glUseProgram(ids[0]);
    }

    public void uniformTexture(GlTexture texture) {
        glActiveTexture(texture.textureUnit);
        glBindTexture(texture.textureType, texture.idTexture);
        int ind = texture.textureUnit - GL_TEXTURE0;
        glUniform1i(samplersLocations[ind], ind);
    }

    public void uniformTexture(int idTexture) {
        int samplerLocation = glGetUniformLocation(idProgram, "sampler");
        glActiveTexture(GL_TEXTURE0);
        glBindTexture(GL_TEXTURE_2D, idTexture);
        glUniform1i(samplerLocation, 0);
    }

    public void uniformMatrix(Matrix4f m) {
        glUniformMatrix4fv(matricesLocations[IND_MATRIX], false, m.get(MATRIX_BUFFER));
    }

    public void uniformSolidColor(float[] color) {
        glUniform4fv(solidColorLocation, color);
    }

    public void uniformMatrices(Matrices m) {
        glUniformMatrix4fv(matricesLocations[IND_MATRIX], false, m.matrix.get(MATRIX_BUFFER));
        glUniformMatrix4fv(matricesLocations[IND_MATRIX_NORMAL], false, m.normals.get(MATRIX_BUFFER));
        glUniformMatrix4fv(matricesLocations[IND_MATRIX_PROJECTION], false, m.projection.get(MATRIX_BUFFER));
        glUniformMatrix4fv(matricesLocations[IND_MATRIX_VIEW], false, m.camera.get(MATRIX_BUFFER));
        glUniformMatrix4fv(matricesLocations[IND_MATRIX_VIEW_MODEL], false, m.viewModel.get(MATRIX_BUFFER));
        glUniformMatrix4fv(matricesLocations[IND_MATRIX_MODEL], false, m.model.get(MATRIX_BUFFER));
    }


    public void uniformDirectionalLight(float[] ambient, float[] lightDiffuse, float[] lightDirection) {
        glUniform3fv(light[0], ambient);
        glUniform3fv(light[1], lightDiffuse);
        glUniform3fv(light[2], lightDirection);
    }

    public void uniformDirectionalLight(Vector3f ambient, Vector3f lightDiffuse, Vector3f lightDirection) {
        glUniform3f(light[0], ambient.x, ambient.y, ambient.z);
        glUniform3f(light[1], lightDiffuse.x, lightDiffuse.y, lightDiffuse.z);
        glUniform3f(light[2], lightDirection.x, lightDirection.y, lightDirection.z);
    }


    public void dispose() {
        glDetachShader(ids[0], ids[1]);
        glDetachShader(ids[0], ids[2]);
        glDeleteProgram(ids[0]);
        glDeleteShader(ids[1]);
        glDeleteShader(ids[2]);
    }

    public void delete() {
        glDetachShader(idProgram, ids[1]);
        glDetachShader(idProgram, ids[2]);
        glDeleteProgram(ids[0]);
    }

    public static int[] createProgram(String vertexShaderSource, String fragmentShaderSource) {
        int idVertexShader = compileShader(GL_VERTEX_SHADER, vertexShaderSource);
        int idFragmentShader = compileShader(GL_FRAGMENT_SHADER, fragmentShaderSource);

        return new int[]{linkProgram(idVertexShader, idFragmentShader), idVertexShader, idFragmentShader};
    }

    public static int linkProgram(int idVertexShader, int idFragmentShader) {
        int idProgram = glCreateProgram();
        glAttachShader(idProgram, idVertexShader);
        glAttachShader(idProgram, idFragmentShader);
        glLinkProgram(idProgram);

        if (glGetProgrami(idProgram, GL_LINK_STATUS) == GL_FALSE) {
            throw new IllegalStateException("Could not link shader program.");
        }
        return idProgram;
    }

    public static int compileShader(int type, String source) {
        int idShader = glCreateShader(type);
        glShaderSource(idShader, source);
        glCompileShader(idShader);

        if (glGetShaderi(idShader, GL_COMPILE_STATUS) == GL_FALSE) {
            String reason = glGetShaderInfoLog(idShader, 500);
            glDeleteShader(idShader);
            throw new IllegalStateException("Could not compile shader: " + reason + "\nSource:\n " + source);
        }
        return idShader;
    }




    public static final String VERTEX_SHADER_120 = "#version 120 \n" +
            "attribute vec4 vCoord;\n" +
            "varying vec4 exColor;\n" +
            "uniform mat4 matrix;\n" +
            "void main() {\n" +
            "    gl_Position = matrix  * vCoord;\n" +
            "    exColor = vec4(1.0,1.0,1.0,1.0);\n" +
            "}";

    public static final String FRAGMENT_SHADER_120 = "#version 120 \n" +
            "varying vec4 exColor;\n" +
            "void main() {\n" +
            "    gl_FragColor = exColor;\n" +
            "}\n";

}

File: lwjglkt/src/main/kotlin/com/darkraha/opengldemokt/gl/ShaderProgram.kt
package com.darkraha.opengldemokt.gl


import com.darkraha.opengldemokt.gl.shader.ShaderProgramBuilder
import com.darkraha.opengldemokt.gl.shader.ShaderProgramBuilder.Companion.U_NORMAL_SAMPLER_NAME
import org.joml.Matrix4f
import org.joml.Vector3f
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL13
import org.lwjgl.opengl.GL20
import org.lwjgl.opengl.GL33.*
import org.lwjgl.system.MemoryUtil


class ShaderProgram {
    private lateinit var ids: IntArray

    val idProgram
        get() = ids[0]

    private val matricesLocations = IntArray(ShaderProgramBuilder.U_MATRIX_NAMES.size)
    private val samplersLocations = IntArray(8)
    private val light = IntArray(3)
    private var normalSamplerLocation = 0

    var solidColorLocation = 0
        private set

    constructor() {
        val builder = ShaderProgramBuilder()
        builder.colors(false, true)
        init(builder.buildVertexShader(), builder.buildFragmentShader())
    }

    constructor(vertexShaderCode: String, fragmentShaderCode: String) {
        init(vertexShaderCode, fragmentShaderCode)
    }

    protected fun init(vertexShaderCode: String, fragmentShaderCode: String) {
        ids = createProgram(vertexShaderCode, fragmentShaderCode)
        setupLocations()
    }


    fun setupLocations() {
        glUseProgram(idProgram)

        for (i in matricesLocations.indices) {
            matricesLocations[i] = glGetUniformLocation(
                idProgram,
                ShaderProgramBuilder.U_MATRIX_NAMES[i]
            )
        }

        samplersLocations[0] = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_SAMPLER_NAME)
        for (i in 1 until samplersLocations.size) {
            samplersLocations[i] = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_SAMPLER_NAME + i)
        }

        solidColorLocation = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_SOLID_COLOR_NAME)

        normalSamplerLocation = glGetUniformLocation(idProgram, U_NORMAL_SAMPLER_NAME)

        light[0] = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_LIGHT_NAMES[0])
        light[1] = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_LIGHT_NAMES[1])
        light[2] = glGetUniformLocation(idProgram, ShaderProgramBuilder.U_LIGHT_NAMES[2])
    }


    fun use() {
        glUseProgram(ids[0])
    }

    fun uniformTexture(texture: GlTexture) {
        glActiveTexture(texture.textureUnit)
        glBindTexture(texture.textureType, texture.idTexture)
        val ind = texture.textureUnit - GL_TEXTURE0
        glUniform1i(samplersLocations[ind], ind)
    }



    fun uniformTexture(idTexture: Int) {
        val samplerLocation = glGetUniformLocation(idProgram, "sampler")
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, idTexture)
        glUniform1i(samplerLocation, 0)
    }

    fun uniformMatrix(m: Matrix4f) {
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX], false,
            m[MATRIX_BUFFER]
        )
    }

    fun uniformSolidColor(color: FloatArray) {
        glUniform4fv(solidColorLocation, color)
    }

    fun uniformMatrices(m: Matrices) {
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX], false,
            m.matrix[MATRIX_BUFFER]
        )
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX_NORMAL], false,
            m.normals[MATRIX_BUFFER]
        )
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX_PROJECTION], false,
            m.projection[MATRIX_BUFFER]
        )
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX_VIEW], false,
            m.view[MATRIX_BUFFER]
        )
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX_VIEW_MODEL], false,
            m.viewModel[MATRIX_BUFFER]
        )
        glUniformMatrix4fv(
            matricesLocations[ShaderProgramBuilder.IND_MATRIX_MODEL], false,
            m.model[MATRIX_BUFFER]
        )
    }


    fun uniformDirectionalLight(ambient: FloatArray, lightDiffuse: FloatArray, lightDirection: FloatArray) {
        glUniform3fv(light[0], ambient)
        glUniform3fv(light[1], lightDiffuse)
        glUniform3fv(light[2], lightDirection)
    }

    fun uniformDirectionalLight(ambient: Vector3f, lightDiffuse: Vector3f, lightDirection: Vector3f) {
        glUniform3f(light[0], ambient.x, ambient.y, ambient.z)
        glUniform3f(light[1], lightDiffuse.x, lightDiffuse.y, lightDiffuse.z)
        glUniform3f(light[2], lightDirection.x, lightDirection.y, lightDirection.z)
    }


    private fun uniformObjectTexture(glObject: GlObject) {
        var usedUnit = 0

        glObject.texture?.also {
            glActiveTexture(GL_TEXTURE0)
            glBindTexture(it.textureType, it.idTexture)
            glUniform1i(samplersLocations[0], 0)
            ++usedUnit
        }

        glObject.normalTexture?.also {
            glActiveTexture(GL_TEXTURE0 + usedUnit)
            glBindTexture(it.textureType, it.idTexture)
            glUniform1i(normalSamplerLocation, usedUnit)
            ++usedUnit
        }

        glObject.extraTextures?.also {
            for (i in it.indices) {
                glActiveTexture(GL_TEXTURE0 + usedUnit)
                glBindTexture(it[i].textureType, it[i].idTexture)
                glUniform1i(samplersLocations[i + 1], usedUnit)
                ++usedUnit
            }
        }
    }

    fun uniform(glObject: GlObject, m: Matrices, light: LightDirectional?) {
        uniformObjectTexture(glObject)

        glObject.transforms?.apply {  m.applyModel(this) }
        uniformMatrices(m)

        light?.apply {
            uniformDirectionalLight(ambient!!, diffuseColor!!, direction!!)
        }
    }


    fun dispose() {
        glDetachShader(ids[0], ids[1])
        glDetachShader(ids[0], ids[2])
        glDeleteProgram(ids[0])
        glDeleteShader(ids[1])
        glDeleteShader(ids[2])
    }

    fun delete() {
        glDetachShader(idProgram, ids[1])
        glDetachShader(idProgram, ids[2])
        glDeleteProgram(ids[0])
    }

    companion object {
        val MATRIX_BUFFER = MemoryUtil.memAllocFloat(16)

        fun createProgram(vertexShaderSource: String, fragmentShaderSource: String): IntArray {
            val idVertexShader: Int = compileShader(GL_VERTEX_SHADER, vertexShaderSource)

            val idFragmentShader: Int =
                compileShader(GL_FRAGMENT_SHADER, fragmentShaderSource)

            return intArrayOf(
                linkProgram(idVertexShader, idFragmentShader),
                idVertexShader,
                idFragmentShader
            )
        }

        fun linkProgram(idVertexShader: Int, idFragmentShader: Int): Int {
            val idProgram = glCreateProgram()
            glAttachShader(idProgram, idVertexShader)
            glAttachShader(idProgram, idFragmentShader)
            glLinkProgram(idProgram)
            check(glGetProgrami(idProgram, GL_LINK_STATUS) != GL_FALSE) { "Could not link shader program." }
            return idProgram
        }

        /**
         * @param type Possible values GL_VERTEX_SHADER, GL_FRAGMENT_SHADER
         */
        fun compileShader(type: Int, source: String): Int {
            val idShader = glCreateShader(type)
            glShaderSource(idShader, source)
            glCompileShader(idShader)
            if (glGetShaderi(idShader, GL_COMPILE_STATUS) == GL_FALSE) {
                val reason = glGetShaderInfoLog(idShader, 500)
                glDeleteShader(idShader)
                throw IllegalStateException("Could not compile shader: $reason\nSource:\n $source")
            }
            return idShader
        }


        @JvmStatic
        val VERTEX_SHADER_120 =
            """
#version 120 
attribute vec4 vCoord;
varying vec4 exColor;
uniform mat4 matrix;
void main() {
    gl_Position = matrix  * vCoord;
    exColor = vec4(1.0,1.0,1.0,1.0);
}
""".trimIndent()

        @JvmStatic
        val FRAGMENT_SHADER_120 =
            """
#version 120 
varying vec4 exColor;
void main() {
    gl_FragColor = exColor;
}
""".trimIndent()
    }
}